### ISSUE TITLE
Postgrex process apparently not starting with Ecto

### ISSUE THREAD

**Issue by henriquesati** [elixir-ecto/ecto#4515]
### Elixir version

26

### Database and Version

Postgres 16

### Ecto Versions

3,6

### Database Adapter and Versions (postgrex, myxql, etc)

 {:postgrex, "~> 0.19.1"},

### Current behavior

I'm trying to connect to Postgres using Ecto and Postgres, using dynamically loaded Repos, but I get the error `exit) shutdown: failed to start child: DBConnection.ConnectionPool
    ** (EXIT) exited in: GenServer.call(DBConnection.Watcher, {:watch, DBConnection.ConnectionPool.Supervisor, {DBConnection.ConnectionPool.Pool, {#PID<0.215.0>, #Reference<0.2546331300.860749830.189489>, Postgrex.Protocol, [types: Postgrex.DefaultTypes, repo: MyApp.Repo, telemetry_prefix: [:my_app, :repo], otp_app: :testes_pay, timeout: 15000, database: "friends", port: 8080, hostname: "localhost", username: "pedri", password: "1234", pool_size: 1, pool: DBConnection.ConnectionPool]}}}, :infinity)
        ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started`
  
I'm being able to use Ecto CLI's such migrations and drops to the db, but when it comes to actually querying it doesnt work
code:
`{:ok, _} = Application.ensure_all_started(:req)
    ct = %{
      chain: "ethereum",
      coin: "ETH",
      transaction_value: 12.345,
      transaction_charged: 1.234,
      ref: "test_ref",
      payload: "test_payload",
      public: "test_public",
      private: "test_private",
      status: false
    }
    
    default_dynamic_repo = MyApp.Repo.get_dynamic_repo()

    {:ok, repo} =
      MyApp.Repo.start_link(
        name: :gatinhos,
        hostname: "localhost",
        username: "pedri",
        password: "1234",
        pool_size: 1
      )
    
    try do
      MyApp.Repo.put_dynamic_repo(repo)
      MyApp.Repo.all(Post)
    after
      MyApp.Repo.put_dynamic_repo(default_dynamic_repo)
      Supervisor.stop(repo)
    end
  
  `
  postgres is running inside a docker on a ubuntu 22 wsl2
   

### Expected behavior

Expected a sucessfull query to the DB

**Comment by josevalim** [comment]
Can you please include your _Elixir_ version (you included the Erlang/OTP one) and try a more recent Ecto just in case? If the issue persists, can you please provide a small snippet that reproduces the error?

**Comment by josevalim** [comment]
Thank you for the added information! Can you condense everything into a repository that you upload to GitHub or a single file with Mix.install that I can run?

**Comment by henriquesati** [comment]
I have no exp with elixir :( but put in a rep https://github.com/henriquesati/ectodb

**Comment by henriquesati** [comment]
hey can you take a look at the rep again? made a commit that make it easier. Just run the  _linkly.ex_, the schema struct its working well, but when I try to registry on the db adding `Repo.insert(user1)` I got  `** (RuntimeError) could not lookup Ecto repo Linkly.Repo because it was not started or it does not exist`
but i'm being able to do migrations and raw sql using iEx

**Comment by henriquesati** [comment]
```
defmodule Linkly.Application do
  # See https://hexdocs.pm/elixir/Application.html
  # for more information on OTP Applications
  @moduledoc false

  use Application

  @impl true
  def start(_type, _args) do
    children = [
      Linkly.Repo,
    ]

    # See https://hexdocs.pm/elixir/Supervisor.html
    # for other strategies and supported options
    opts = [strategy: :one_for_one, name: Linkly.Supervisor]
    Supervisor.start_link(children, opts)
  end
end

```

**Comment by josevalim** [comment]
I see. The issue is that you are giving it a name of `nil`. Because of that, its process cannot be found, because it has no name. If you want to run scripts within your Elixir app, you must do `mix run path/to/file`. For IEx, you need to do `iex -S mix`.

**Comment by henriquesati** [comment]
Ecto docs says this: `Ecto also allows you to start a repository with no name (nil) (just like that famous horse). In such cases, you need to explicitly pass name: nil and match on the result of MyApp.Repo.start_link/1 to retrieve the PID, which should be given to put_dynamic_repo. Let's also use this opportunity and perform proper database clean-up, by shutting up the new repository and reverting the value of put_dynamic_repo:` [aqui](https://hexdocs.pm/ecto/replicas-and-dynamic-repositories.html)

besides that I still dont know why my I cant connect to db using the static repo

**Comment by josevalim** [comment]
Yes, but the issue is that main repository is not started either. Move the code from outside of `lib/linkly.ex` to the root of the project as `linkly.exs` and run it as `mix run linkly.exs`.

**Comment by henriquesati** [comment]
moving Linkly.exs to the root worked at least with the dynamic repo, can you explain why? what would be that main repository?

**Comment by josevalim** [comment]
Files in `lib` are compiled. You should not be accessing the database during compilation. With `mix run`, your app is compiled and its repository is started, so you can use it regularly.
